Implement retry spec in task spec and retry logic in the workflow conductor

userlocalhost opened the pull request:
# Abstract
I added an implementation for the feature of retry which is mentioned in #118 by making the use of the nick's work (c.f. #171).

# Changing points
The are the main changing points of implementation for retry feature in Conductor.

* Added parameters for task_state_entry to keep these information
  - retry.count: how many times the target task was executed already
  - retry.running_time: how long target task should wait for retrying next time when 'delay' parameter was specified

* Added processing that makes task scheduling delayed when it is ordered to be waited in the get_next_tasks method.

* Added processing to select whether target task would be retry or not in the update_task_state method.

## Note
This doesn't implement feature of infinity retry processing intensionally because this is middle of controversial discussion (c.f. https://github.com/StackStorm/orquesta/issues/118#issuecomment-513106400).

# Execution results
Here is an example action definition to show how it works.
### metadata
```YAML
---
name: retry_test
pack: default
description: A workflow for testing retry feature of orquesta
runner_type: orquesta
entry_point: workflows/retry_test.yaml
enabled: true
parameters:
  count:
    type: integer
    default: 1
  delay:
    type: integer
    default: 0
  msg:
    type: string
    default: ""
```

### workflow
```YAML
version: 1.0

input:
  - count
  - delay
  - msg

tasks:
  task:
    action: slack.post_message
    input:
      message: '[<% now().format("%Y-%m-%d %H:%M:%S") %>] <% ctx(msg) %>'
    retry:
      when: <% succeeded() %>
      count: <% ctx(count) %>
      delay: <% ctx(delay) %>
````

They are execution results, left-side is run above action without delay parameter and right-side is another one with delay parameter. They were run almost at the same time, and we could confirm that task would be retried for specified times at 'count' parameter and delayed to be scheduled for specified seconds at 'delay' parameter.

<img width="1250" alt="スクリーンショット 2019-10-02 15 42 52" src="https://user-images.githubusercontent.com/469934/66023192-61e0fd00-e52b-11e9-90e0-7240ac994dfa.png">
<img width="597" alt="スクリーンショット 2019-10-02 14 57 25" src="https://user-images.githubusercontent.com/469934/66023104-1cbccb00-e52b-11e9-8c8d-66d70768c353.png">

armab commented:
Great stuff, thanks for contribution and effort to continue this highly useful feature implementation!

I'll let @m4dcoder to take a look at it closer.

m4dcoder commented:
@nmaludy Are you and @userlocalhost working together on the Retry spec?  There's already a WIP PR for Retry at https://github.com/StackStorm/orquesta/pull/171.  That PR is also mentioned here.

userlocalhost commented:
Thank you for confirming that. This PR inherits the change of #171 (which adds spec for Retry to TaskSpec). I added an implementation of its feature in the conductor processing.
Sorry for making you confused, because I made another PR for same feature. I guess I should have made a PR to nick's branch not to make another one.

Before starting this work, I consulted nick about whether I can take over the rest of his work in the #community channel. And I engaged in this work by getting his consent.
<img width="662" alt="スクリーンショット 2019-10-09 10 41 39" src="https://user-images.githubusercontent.com/469934/66445349-710f0000-ea81-11e9-91df-322955bf3b05.png">

Would you please review it, @m4docker?

m4dcoder commented:
@userlocalhost Got it! Thank you for picking up this work. I did review this. I will post my comments on this shortly today.

m4dcoder commented on orquesta/conducting.py:
```diff
@@ -730,6 +761,50 @@ def update_task_state(self, task_id, route, event):
+                            task_id,
+                            route,
+                            ctxs=task_state_entry['ctxs']['in'],
+                            prev=task_state_entry['prev'],
```
Should probably include the `delay` here so it can be written to the staged task, which will be passed to StackStorm for processing.

m4dcoder commented on orquesta/conducting.py:
```diff
@@ -730,6 +761,50 @@ def update_task_state(self, task_id, route, event):
+                            task_state_entry['ctxs']['retry']['running_time'] = (
+                                int(datetime.datetime.now().strftime('%s')) +
+                                int(expr_base.evaluate(task_spec.retry.delay, current_ctx))
+                            )
```
No need to set the timer here.  See comments above on sleep.

m4dcoder commented on orquesta/conducting.py:
```diff
@@ -730,6 +761,50 @@ def update_task_state(self, task_id, route, event):
 
+            # This decides whether this task would be retried or not when 'retry' is specified
+            may_retry = (
+                hasattr(task_spec, 'retry') and
```
This is not necessary.  The `task_spec.has_retry` does the same.

m4dcoder commented on orquesta/conducting.py:
```diff
@@ -730,6 +761,50 @@ def update_task_state(self, task_id, route, event):
+                hasattr(task_spec.retry, 'when') and
+                task_spec.has_retry()
+            )
+            if may_retry:
```
Instead of doing may_retry and then will_retry, can we check all of them in a single if statement?  I prefer this over multiple indented blocks of codes.
`if task_spec.has_retry() and if number of retry has exhausted and if retry when condition is satisfied:`

m4dcoder commented on orquesta/conducting.py:
```diff
@@ -730,6 +761,50 @@ def update_task_state(self, task_id, route, event):
+                        task_state_entry['status'] = new_task_status = old_task_status
+
+                        # Increment task count to keep track of how many times task is repeated.
+                        task_state_entry['ctxs']['retry']['count'] += 1
```
See comments above. Write to `task_state_entry['retry']['count'] += 1` instead

m4dcoder commented on orquesta/conducting.py:
```diff
@@ -730,6 +761,50 @@ def update_task_state(self, task_id, route, event):
+                    )
+                    if will_retry:
+                        # This turns task_status back to old one.
+                        task_state_entry['status'] = new_task_status = old_task_status
```
I am struggling with this. The state machine `machines.TaskStateMachine.process_event` at line https://github.com/StackStorm/orquesta/blob/master/orquesta/conducting.py#L706 should be the one to determine the task status.  This is basically going outside of the state machine to make changes to the task status. Is it possible to let the TaskStateMachine set the new task status to `retrying` and the `if may_retry:` above is just `if new_task_status == statuses.RETRYING:`?

userlocalhost replied:
I wanted to process task-retry by using [the state transition machine for task](https://github.com/StackStorm/orquesta/blob/master/orquesta/machines.py#L551). But it's difficult to implement because retrying processing is non deterministic. Therefore, I implemented it to rollback task status by using evaluator.

---

What makes it difficult to do it is state transition is determined by following three factors.
  - current task's status
  - current taks's retry-count
  - user defined condition which is specified in 'when' parameter in metadata

When I define task transitions for each status to take into account the retry-count and user-specified conditions, it might be possible to implement this state transition processing by using TASK_STATE_MACHINE_DATA.

I also believe that it's more elegant implementation. But It's hard for me to implement it without omission and bugs. And it's easy to implement the processing to determine whether task should be retried by using evaluator. That takes into account both current status and user specified condition to retry at the same time.

So please forgive me to implement that feature to rollback status after changing status when state-machine determines task will be retried.

userlocalhost commented:
Thank you very much for your review in detail. All of the points that I should correct are clear by your comments.

About the changes in the `update_task_state` method, I will make this change simple by taking into account a new state transition for retry.

And for the processing of delay, I should be able to offload this processing to workflow-engine or action runner.
I didn't understand what each processes and components should dedicate in until I got your review. I could understand that Orquesta is just "controller" to indicate what task(s) should be run next, to the "processor" which dedicate to run it. So, the processing to wait until the time is not entitle to set on the controller, but the processor.

But, I'm not sure exactly where that delay processing should be set in the processor yet, I think this should be set on somewhere before running action in an action runner process, so far.

I was wondering if you could give me some advice about it after fixing these changes.

m4dcoder commented:
If I understand you correctly, you don't need to worry how StackStorm process the delay.  This is already implemented.  When you setup the task dict for staging, you just need to add a KVP for delay and make sure the task dict returned from get_next_tasks has the delay key set.

userlocalhost commented:
Thank you for giving important advice. I could understand where I should fix on orquesta for making delay by seeing the implementation of [request_task_execution](https://github.com/StackStorm/st2/blob/master/st2common/st2common/services/workflows.py#L433).

I plan on dealing with it within a few days.

userlocalhost commented:
The last commit refactored following points.

- Eliminated sleep processing and specified `delay` parameter to order workflow engine to make task delayed.
- Encapsulated task-retry processing to TaskState instance, to be able to retry task transparently by calling update_task_state method without changing that implementation.

And following is an execution results of refactored processing. I could confirm that both delay and count parameters are worked.

<img width="1207" alt="スクリーンショット 2019-10-21 0 10 44" src="https://user-images.githubusercontent.com/469934/67173487-ea9bdc00-f3f9-11e9-835c-ca81dbe6703e.png">
<img width="409" alt="スクリーンショット 2019-10-21 0 10 54" src="https://user-images.githubusercontent.com/469934/67173488-ea9bdc00-f3f9-11e9-82be-927fd000a2b4.png">

userlocalhost commented:
I'll add some changes to increase code coverage of this change, so I amended PR title as 'WIP'.

userlocalhost commented:
I fixed what you mentioned. Would you please review it, @m4dcoder ?

m4dcoder commented:
@userlocalhost Yeah, I see the changes. Thank you. I'll review it sometime this week.

m4dcoder commented:
@userlocalhost I want to let you know that I started reviewing this PR. There is some nice idea with TaskState. But there are also concerns.  I'm still thinking through this so please be patient. Thanks.

userlocalhost commented:
@m4dcoder Thank you for your review. Don't worry about it. I'm happy to fix!

userlocalhost commented:
Thank you very much for your minute review, @m4dcoder.

I could understand what I should fix, and also realized my previous fixing was halfway measures to implement this feature.
I shouldn't have focused on being less change of current implementation at the expense of maintainability.

And I’m sorry for taking up your time because of my inadequate implementation which doesn't act on principle of task state machine (especially, the change to alter task state outside of [the task state transition processing using WORKFLOW_STATE_MACHINE_DATA](https://github.com/StackStorm/orquesta/blob/master/orquesta/machines.py#L710)).
I'll fix it to be careful of keeping this consistent system.

userlocalhost commented on orquesta/conducting.py:
```diff
@@ -476,6 +573,11 @@ def get_task(self, task_id, route):
+        # When task is retried, orquesta order to delay task execution for specified seconds
+        if (task_spec.has_retry() and task_spec.retry.delay and
+                task_state and task_state.is_retried()):
+            task['delay'] = expr_base.evaluate(task_spec.retry.delay, task_ctx)
```
I understood that processing should be written in the proper place by your comment. In this case, `get_task` is the function that dedicates to get task that should be run in the workflow engine. And processing of attaching additional information (e.g. delaying time) to it is not eligible to write in here.

So I move this processing to `_evaluate_task_actions` method that do these processing.

userlocalhost commented:
I refactored the implementation of task retry feature. Would you please review it again, @m4dcoder ?

Following are major changing points from last commit.

#### 1. Reverted data-structure of task_state_entry as dict from TaskState class.

Followings are reasons of this change.

  - Classification of this data-structure is not essential of this feature. At first time, I thought that classifying attitudes of task_state would increase the maintainability of this code. But I realized that feature would be realized by just adding few related attribute.
  - Classification of this data-structure requires big effort. This is valuable for increasing maintainability and expandability of this code, I think. But this change may cover a lot of ground. I want to tackle of this work as another task.

#### 2. Implemented the processing to change status of task using state transition system of TaskStateMachine. 

  Previous implementation I made change task status by attached processing after state transition. It was foreign matter that should be get rid of, because it didn't act on the principle of state transition in the orquesta.

#### 3. Making a restriction for when parameter of retry spec, that parameter accepts only either "succeeded" or "failed".

  Mostly, user specifies `when: <% failed() %>` for retrying task when it is failed, I think. But it has implementing constraint, that both `<% succeeded() %>` and `<% failed() %>` expressions are evaluated as expected on the basis of the status after changing. In other words, orquesta can't determine whether task should be retried before task is failed or succeeded as long as using this expression.

  By this constraint, I had no choice but to implement retrying task after task was transitioned (but this is a bad choice as mentioned above). This is because the evaluation results of these expressions are determined dynamically. On the other hand, condition of retrying task can be specified statically (e.g. retrying task when it is failed (or succeeded)).

  So, I changed TaskRetrySpec to accept only statical condition (`failed` or `succeeded`). By this change, we are free from that implementing constraint and the implementation of retrying feature would be simple.

userlocalhost commented:
Sorry I realized the lack of consideration about the retry feature when the workflow was abandoned.

Once some problems were happened at a workflow processing, orquesta would change status of workflow and all tasks to `statuses.FAILED`. But in the PR implementation, task's status would be still "running" when retry condition would be satisfied.

I should be able to handle in this case, so I changed PR title as WIP.

userlocalhost commented:
I changed following in addition to https://github.com/StackStorm/orquesta/pull/180#issuecomment-550905638, after the last review.

Once again, would you please review this, @m4dcoder ?

---

#### 4. Added a function and condition to prevent retrying task even though its condition was satisfied when workflow was aborted.

I checked the implementation of request_workflow_status carefully abount whether a task which has retry statement would be retried when workflow is aborted like [here](https://github.com/StackStorm/orquesta/blob/master/orquesta/conducting.py#L429). I understood that calling `conductor.request_workflow_status(statuses.FAILED)` would never affect status of each tasks.

And I also think, task should not be retried any more after the workflow is aborted. So I added a function to exhaust retry_count not to do it, and condition not to set valid retry_count in this case.

m4dcoder commented:
@userlocalhost Ack. I will review your changes this week. Thank you for your diligence here. This is a much requested feature that the community will appreciate.

userlocalhost commented:
Sorry, I realized that I took serious mistake about the spec of inflexible retry condition. This PR wouldn't be satisfied with functional requirement and consistency.

Originally, orquesta allows to specify flexible condition of task transition as is mentioned https://github.com/StackStorm/orquesta/issues/118#issuecomment-457400742.
That indicates that retry condition also should be able to specify flexible one. 

But the following implementation (c.f. https://github.com/StackStorm/orquesta/pull/180#issuecomment-550905638) of my last commit is adverse to it.

> 3. Making a restriction for when parameter of retry spec, that parameter accepts only either "succeeded" or "failed".

Sorry for asking for review with wrong implementation again and again.
I should be able to fix it comply with task transition principle in orquesta this time.

m4dcoder commented:
@userlocalhost Yeah, I saw that. No worries. I've been reviewing and I'm currently refactoring the PR since we've been back and forth with this PR for awhile.

userlocalhost commented:
[The last commit](https://github.com/StackStorm/orquesta/pull/180/commits/5cbf58b2cf8d963743e170e7902e8273451b9c5a) fixes TaskRetrySpec and the implementation of TaskStateMachine to be able to specify flexible retry condition as well as being consistent with other specs.

That is realized by following changes.

- Fixed input spec of 'when' parameter in the TaskRetrySpec, which describes task retry condition, to be able to specify flexible condition that is same as others (e.g. [TaskTransitionSpec](https://github.com/StackStorm/orquesta/blob/master/orquesta/specs/native/v1/models.py#L48)).
- Refactored processing for task retry in the TaskStateMachine. The fixed processing recalls the method of TaskStateMachine to transit task state and attach context of retrying when task state and results match with retry condition.

Would you please review this, @m4dcoder?

userlocalhost commented:
Just in case, I show an execution result of following workflow using the orquesta which is deployed by last commit in this PR branch (That is NOT different from previous ones).

```
version: 1.0

input:
  - count
  - delay
  - msg

tasks:
  task:
    action: slack.post_message
    input:
      message: '[<% now().format("%Y-%m-%d %H:%M:%S") %>] <% ctx(msg) %>'
    retry:
      when: <% succeeded() %>
      count: <% ctx(count) %>
      delay: <% ctx(delay) %>
```
<img width="1242" alt="スクリーンショット 2019-11-25 12 16 44" src="https://user-images.githubusercontent.com/469934/69511228-c300dc00-0f82-11ea-9d1f-45d01a35f063.png">
<img width="474" alt="スクリーンショット 2019-11-25 12 14 29" src="https://user-images.githubusercontent.com/469934/69511227-c2684580-0f82-11ea-94c2-b0c854f67bfa.png">

m4dcoder commented:
@userlocalhost After review this PR a few times, I decided to take the responsibility to refactor the PR here. Otherwise, we would still be going back and forth with the changes here. I commend you on your previous attempts. There are multiple ways to implement the retry in the workflow conductor.  This latest iteration is just a representation of what I believe is the simplest approach and staying consistent with the current design within minimal changes to interfaces.

The workflow conductor here is responsible for evaluating whether a task needs to be retried. The retry evaluation takes place only after the new task status is determined by the task state machine. This is required because the retry `when` expression is generic and is not known at design time of the conductor and the expression may look at the task status and/or task result to determine if a retry is required. When a retry is required, the conductor will post a task retry event to the task state machine which will set the task status to `retrying`, which then tally the retry and sets up the task to be returned by `get_next_tasks`.

Another consideration why I made the change here is that our previous thought about letting the state machines make the decision didn't quite work out as we expected. In the last iteration, we end up having to pass the `will_retry` decision explicitly anyway from the conductor and the conductor has to look at the returned altered event to determine if a retry is required.

Anyway, this retry feature is not a trivial change and requires some experimentation.  I appreciate your efforts here. Thank you for your hard work.

userlocalhost commented:
@m4dcoder 
Thank you very much for your help. And I apologize I couldn't come up to your expectation.

I feel a little bit regretful at myself not to be able to finish up this work for myself, but I also feel excited because it was very good experience for me, that I could learn about internal of orquesta and interaction with st2workflow.

And I believe my understanding about the philosophy of orquesta would be better by catching up with your work.

Finally, I really appreciate your comments. Your advice always makes me motivated. I couldn't have come this far without your support.
I continue to be involved with this community. And I hope to be able to make a contribution someday in the future.

Thanks,

m4dcoder commented:
@userlocalhost We appreciate user feedback and contribution.  There's no expectation that needed to be met. This feature requires trial and error and experimentation. We are trying to come up with the best approach. Your previous attempts helped us quite a lot and tells us what we can or cannot do.  It's a job well done.  I just didn't want to drag this PR for too long and discourage you.  So I'm glad to hear that you will continue to look for the next opportunity to contribute.

m4dcoder commented:
Required StackStorm update and integration tests at https://github.com/StackStorm/st2/pull/4833.
